lib/sycamore/sycamore/connectors/neo4j/neo4j_writer.py

```diff
@@ -0,0 +1,343 @@
+            #### merge nodes together ####
+            for key, values in headers2["nodes"].items():
+                if headers1["nodes"].get(key, None) is None:
+                    headers1["nodes"][key] = values
+                else:
+                    for value in values:
+                        headers1["nodes"][key][value] = True
+            #### merge relationships together ####
+            for key, values in headers2["relationships"].items():
+                if headers1["relationships"].get(key, None) is None:
+                    headers1["relationships"][key] = values
+                else:
+                    for value in values:
+                        headers1["relationships"][key][value] = True
+            return headers1
+
+        def finalize(nodes):
+            return nodes
+
+        aggregation = AggregateFn(
+            init=lambda group_key: dict(nodes=dict(), relationships=dict()),
+            accumulate_row=accumulate_row,
+            merge=merge,
+            finalize=finalize,
+            name="output",
+        )
+
+        return self._dataset.aggregate(aggregation)["output"]
+
+    def _write_nodes_csv_headers(self, nodes):
+        for node_label, node_columns in nodes.items():
+            csv_path = os.path.expanduser(f"{self._import_dir}/sycamore/nodes/{node_label}.csv")
+            headers = sorted(set(column for column in node_columns.keys()))
+            os.makedirs(os.path.dirname(csv_path), exist_ok=True)
+            with open(csv_path, "w", newline="") as file:
+                writer = csv.writer(file)
+                writer.writerow(headers)
+
+    def _write_relationships_csv_headers(self, relationships):
+        for relationship_label, relationship_columns in relationships.items():
+            csv_path = os.path.expanduser(f"{self._import_dir}/sycamore/relationships/{relationship_label}.csv")
+            headers = sorted(set(column for column in relationship_columns.keys()))
+            os.makedirs(os.path.dirname(csv_path), exist_ok=True)
+            with open(csv_path, "w", newline="") as file:
+                writer = csv.writer(file)
+                writer.writerow(headers)
+
+
+class Neo4jWriteCSV(MapBatch, Write):
+    def __init__(self, plan: Node, client_params: Neo4jWriterClientParams, **kwargs):
+        self._import_dir = client_params.import_dir
+        self._name = "Neo4jWriter"
+        super().__init__(plan, f=self._write_docs_tt, **kwargs)
+
+    def _parse_docs(self, docs):
+        include = ["type", "bbox", "text_representation"]
+        nodes = defaultdict(list)
+        relationships = defaultdict(lambda: defaultdict(list))
+        for doc in docs:
+            if "label" not in doc.data or "doc_id" not in doc.data:
+                continue
+            node = {
+                "uuid:ID": doc.doc_id,
+            }
+            if doc["type"] == "table":
+                doc["text_representation"] = doc["table"].to_csv()
```

Comment:
sometimes tables are None, so this is a little spicy. Also there's a key ('_override_text') that gets used with TableElements when you set the text_representation (to something other than table.to_csv() - see the implementation of TableElement.text_representation to see how that works and how to make this work with that. You can also change that implementation if you have a better way to get the same behavior)

Reply:
okay after looking at implementation, it will work fine without the table check at all. removed

Reply:
take it back. does not work. So whats happening is that this is actually no longer an element. To represent it in a hierarchical way, I took the element data and stuck it into a hierarchical document. I would like to extract this away from the user though, so I'll make a change in the hierarchical document for now that supports this functionality. Worth a conversation on the data model later on though.